### PR TITLE
react-shop(dev)

### PR DESCRIPTION
**Pull request by VladSobko**

### CODE REVIEW

File: src/LoginPage.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                type="submit"
+                className="btn btn-primary"
+                disabled={isSubmitting}
+              >
+                Login
+              </button>
+              {isSubmitting && (
+                <img
+                  src="data:image/gif;base64,R0lGODlhEAAQAPIAAP///wAAAMLCwkJCQgAAAGJiYoKCgpKSkiH/C05FVFNDQVBFMi4wAwEAAAAh/hpDcmVhdGVkIHdpdGggYWpheGxvYWQuaW5mbwAh+QQJCgAAACwAAAAAEAAQAAADMwi63P4wyklrE2MIOggZnAdOmGYJRbExwroUmcG2LmDEwnHQLVsYOd2mBzkYDAdKa+dIAAAh+QQJCgAAACwAAAAAEAAQAAADNAi63P5OjCEgG4QMu7DmikRxQlFUYDEZIGBMRVsaqHwctXXf7WEYB4Ag1xjihkMZsiUkKhIAIfkECQoAAAAsAAAAABAAEAAAAzYIujIjK8pByJDMlFYvBoVjHA70GU7xSUJhmKtwHPAKzLO9HMaoKwJZ7Rf8AYPDDzKpZBqfvwQAIfkECQoAAAAsAAAAABAAEAAAAzMIumIlK8oyhpHsnFZfhYumCYUhDAQxRIdhHBGqRoKw0R8DYlJd8z0fMDgsGo/IpHI5TAAAIfkECQoAAAAsAAAAABAAEAAAAzIIunInK0rnZBTwGPNMgQwmdsNgXGJUlIWEuR5oWUIpz8pAEAMe6TwfwyYsGo/IpFKSAAAh+QQJCgAAACwAAAAAEAAQAAADMwi6IMKQORfjdOe82p4wGccc4CEuQradylesojEMBgsUc2G7sDX3lQGBMLAJibufbSlKAAAh+QQJCgAAACwAAAAAEAAQAAADMgi63P7wCRHZnFVdmgHu2nFwlWCI3WGc3TSWhUFGxTAUkGCbtgENBMJAEJsxgMLWzpEAACH5BAkKAAAALAAAAAAQABAAAAMyCLrc/jDKSatlQtScKdceCAjDII7HcQ4EMTCpyrCuUBjCYRgHVtqlAiB1YhiCnlsRkAAAOwAAAAAAAAAAAA=="
```
Comment (Di4ik3r) [inline src/LoginPage.js]: думаю це можна винести кудись: в окрему константу, або десь вище

File: src/components/Admin/AdminPage.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+//                   Add New
+//                 </button>
+//               </div>
+//             </div>
+//           </div>
+//         </div>
+//       </div>
+//     );
+//   }
+// }
```
Comment (Di4ik3r) [inline src/components/Admin/AdminPage.js]: чи є необхідність цього комента?

File: src/components/Admin/AdminPage.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    dispatch
+  );
+};
+// const mapDispatchToProps = dispatch => {
+//   return {
+//     addProduct: id => dispatch(addProduct(id)),
+//     deleteProduct: id => dispatch(deleteProduct(id)),
+//     updateProduct: id => dispatch(updateProduct(id)),
+//   };
+// };
```
Comment (Di4ik3r) [inline src/components/Admin/AdminPage.js]: те ж саме, чи є необхідність в цьому коменті?

File: src/components/Admin/ProductItem.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            Edit
+          </button>
+        </td>
+        <td>
+          <button className="btn btn-danger" onClick={this.deleteProduct}>
+            Delete
+          </button>
+        </td>
+      </tr>
+    );
```
Comment (Di4ik3r) [inline src/components/Admin/ProductItem.js]: було б зручніше винести в окремі компоненти

File: src/components/Cart/Cart.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import Navbar from "../Navbar/Navbar";
+
+const Cart = ({ cart }) => {
+  const [totalPrice, setTotalPrice] = useState(0);
+  const [totalItems, setTotalItems] = useState(0);
+
+  useEffect(() => {
+    let items = 0;
+    let price = 0;
+    //JSON.parse(localStorage.getItem("products")).cart.forEach((item)
```
Comment (Di4ik3r) [inline src/components/Cart/Cart.js]: комент

File: src/components/Navbar/Navbar.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import styles from "./Navbar.module.css";
+
+import { connect } from "react-redux";
+
+const Navbar = ({ cart }) => {
+  const [cartCount, setCartCount] = useState(0);
+
+  useEffect(() => {
+    let count = 0;
+    // JSON.parse(localStorage.getItem("products")).
```
Comment (Di4ik3r) [inline src/components/Navbar/Navbar.js]: комент

File: src/redux/Shopping/shopping-reducer.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+      "These books will keep you busy all throughout the entire lockdown and give you some great advise from famous people",
+    price: 150.0,
+    image:
+      "https://images.unsplash.com/photo-1512820790803-83ca734da794?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1374&q=80",
+  },
+];
+
+const INITIAL_STATE = {
+  products:
+    // JSON.parse(localStorage.getItem("products")).products ||
```
Comment (Di4ik3r) [inline src/redux/Shopping/shopping-reducer.js]: комент

File: src/redux/Shopping/shopping-reducer.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      );
+      let stateCopyOnAddToCart = {
+        ...state,
+        cart: inCart
+          ? state.cart.map((item) =>
+              item.id === action.payload.id
+                ? { ...item, qty: item.qty + 1 }
+                : item
+            )
+          : [...state.cart, { ...item, qty: 1 }],
```
Comment (Di4ik3r) [inline src/redux/Shopping/shopping-reducer.js]: спробуй винести цю трьохетажну тернарку по змінним

File: src/LoginPage.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        render={({ errors, status, touched, isSubmitting }) => (
+          <Form>
+            <div className="form-group">
+              <label htmlFor="username">Username</label>
+              <Field
+                name="username"
+                type="text"
+                className={
+                  "form-control" +
+                  (errors.username && touched.username ? " is-invalid" : "")
```
Comment (Di4ik3r) [inline src/LoginPage.js:42]: `is-invalid` краще виносити в константи

File: src/LoginPage.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        <br />
+        <strong>Administrator</strong> - U: admin P: admin
+      </div>
+      <h2>Login</h2>
+      <Formik
+        initialValues={{
+          username: "",
+          password: ""
+        }}
+        validationSchema={Yup.object().shape({
```
Comment (Tymofiev) [inline src/LoginPage.js]: move to validation schemas file

File: src/LoginPage.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+          username: Yup.string().required("Username is required"),
+          password: Yup.string().required("Password is required")
+        })}
+        onSubmit={({ username, password }) => {
+          authenticationService.login(username, password)
```
Comment (Tymofiev) [inline src/LoginPage.js]: more readable would be above render create `handleLogin` function

File: src/LoginPage.js (hunk 1 above)
Comment (Tymofiev) [reply src/LoginPage.js]: +1

File: src/components/Admin/AdminPage.js [hunk 12]
```diff
@@ -0,0 +1,185 @@
+import React, { useCallback } from "react";
+
+import ProductList from "./ProductList";
```
Comment (Tymofiev) [inline src/components/Admin/AdminPage.js]: reorder imports => npm packages first

File: src/components/Admin/AdminPage.js [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+              </button>
+            </div>
+          </div>
+        </div>
+      </div>
+    </div>
+  );
+};
+
+// class AdminPage extends Component {
```
Comment (Tymofiev) [inline src/components/Admin/AdminPage.js]: remove comment

File: src/components/Admin/ProductItem.js [hunk 14]
```diff
@@ -0,0 +1,90 @@
+import React, { Component } from "react";
+
+export default class ProductItem extends Component {
+  constructor(props) {
+    super(props);
+    this.state = { isEdit: false };
+    this.editProduct = this.editProduct.bind(this);
+    this.editProductSubmit = this.editProductSubmit.bind(this);
+    this.deleteProduct = this.deleteProduct.bind(this);
+  }
```
Comment (Tymofiev) [inline src/components/Admin/ProductItem.js:44]: add empty spaces to split code sections

File: src/components/Admin/ProductItem.js [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+      this.props.product.id,
+      this.titleInput.value,
+      this.descriptionInput.value,
+      this.priceInput.value,
+      this.imageInput.value
+    );
+  }
+  render() {
+    const { title, description, price, image } = this.props.product;
+    return this.state.isEdit === true ? (
```
Comment (Tymofiev) [inline src/components/Admin/ProductItem.js]: why explicit comparing? i would prefer to use `const { isEdit } = this.state` => isEdit ? ... : ..

File: src/components/Admin/ProductList.js [hunk 16]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import ProductItem from "./ProductItem";
+
+const ProductList = ({ productList, editProductSubmit, deleteProduct }) => {
+  let products = productList;
```
Comment (Tymofiev) [inline src/components/Admin/ProductList.js]: looks useless to me, if afraid of mutation by ref use distruction `[...productList].map`

File: src/components/Admin/ProductList.js [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+  const trItem = products.map((item, index) => (
```
Comment (Tymofiev) [inline src/components/Admin/ProductList.js]: i would not create var for this => directly to render

File: src/components/Admin/ProductList.js [hunk 18]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import ProductItem from "./ProductItem";
```
Comment (Tymofiev) [inline src/components/Admin/ProductList.js:3]: split own modules from npm

File: src/components/Admin/ProductItem.js (hunk 4 above)
Comment (Tymofiev) [reply src/components/Admin/ProductItem.js]: +1

File: src/components/Admin/ProductItem.js [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+  editProductSubmit() {
+    this.setState((prevState, props) => ({
+      isEdit: !prevState.isEdit,
+    }));
+    this.props.editProductSubmit(
+      this.props.product.id,
+      this.titleInput.value,
+      this.descriptionInput.value,
+      this.priceInput.value,
+      this.imageInput.value
```
Comment (Tymofiev) [inline src/components/Admin/ProductItem.js]: why dont use the form instead of refs?

File: src/components/Admin/ProductItem.js [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+          <button className="btn btn-primary" onClick={this.editProductSubmit}>
+            Save
+          </button>
+        </td>
+        <td>
+          <button className="btn btn-danger">Delete</button>
+        </td>
+      </tr>
+    ) : (
+      <tr key={this.props.index}>
```
Comment (Tymofiev) [inline src/components/Admin/ProductItem.js]: use one code style either distructuring props or straight through `this.prop`
first option fits better as for me

File: src/components/Cart/Cart.js [hunk 21]
```diff
@@ -0,0 +1,57 @@
+import React, { useEffect, useState } from "react";
+import styles from "./Cart.module.css";
```
Comment (Tymofiev) [inline src/components/Cart/Cart.js]: import order

File: src/components/Cart/Cart.js [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+
+const Cart = ({ cart }) => {
+  const [totalPrice, setTotalPrice] = useState(0);
+  const [totalItems, setTotalItems] = useState(0);
+
+  useEffect(() => {
+    let items = 0;
+    let price = 0;
+    //JSON.parse(localStorage.getItem("products")).cart.forEach((item)
+    cart.forEach((item) => {
```
Comment (Tymofiev) [inline src/components/Cart/Cart.js]: try rewrite using reduce and without creating 2 let vars, more elegant and readable

File: src/components/Cart/Cart.js [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+    setTotalPrice(price);
+  }, [cart, totalPrice, totalItems]);
+
+  return (
+    <>
+      <Navbar />
+      <div className={styles.cart}>
+        <div className={styles.cart__items}>
+          {/* cart */}
+          {JSON.parse(localStorage.getItem("products")).cart.map((item) => (
```
Comment (Tymofiev) [inline src/components/Cart/Cart.js]: 1. never can be so sure of `cart` being there
2. move from render to var above

File: src/components/Navbar/Navbar.js [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+    cart.forEach((item) => {
```
Comment (Tymofiev) [inline src/components/Navbar/Navbar.js]: reduce

File: src/components/SingleItem/SingleItem.js [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+          >
+            Add To Cart
+          </button>
+        </div>
+      </div>
+    </>
+  );
+};
+
+const mapStateToProps = state => {
```
Comment (Tymofiev) [inline src/components/SingleItem/SingleItem.js:38]: if you don't need dispatch to be passed into props, use short form straight into `connect((state) => ({ currItem: state... }))` wrap

File: src/helpers/fake-backend.js [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+        );
+        if (!user) return error("Username or password is incorrect");
+        return ok({
+          id: user.id,
+          username: user.username,
+          firstName: user.firstName,
+          lastName: user.lastName,
+          role: user.role,
+          token: `fake-jwt-token.${user.role}`,
+        });
```
Comment (Tymofiev) [inline src/helpers/fake-backend.js]: empty spaces would be great

File: src/helpers/fake-backend.js [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+          (x) =>
+            x.username === params.username && x.password === params.password
+        );
+        if (!user) return error("Username or password is incorrect");
+        return ok({
+          id: user.id,
+          username: user.username,
+          firstName: user.firstName,
+          lastName: user.lastName,
+          role: user.role,
```
Comment (Tymofiev) [inline src/helpers/fake-backend.js]: distructure using spread operator

File: src/redux/Shopping/shopping-actions.js [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+export const loadCurrentItem = (item) => {
+  return {
+    type: actionTypes.LOAD_CURRENT_ITEM,
+    payload: item,
+  };
+};
+
+export function addProduct(product) {
+  return {
+    type: "ADD_PRODUCT",
```
Comment (Tymofiev) [inline src/redux/Shopping/shopping-actions.js]: always move const to file or at least separate var

File: src/redux/Shopping/shopping-reducer.js [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+const shopReducer = (state = INITIAL_STATE, action) => {
+  switch (action.type) {
+    case actionTypes.ADD_TO_CART:
+      // Get Item data from products array
+      const item = state.products.find(
+        (product) => product.id === action.payload.id
+      );
+      // Check if Item is in cart already
+      const inCart = state.cart.find((item) =>
+        item.id === action.payload.id ? true : false
```
Comment (Tymofiev) [inline src/redux/Shopping/shopping-reducer.js]: `item.id === action.payload.id` is already a bool

File: src/redux/Shopping/shopping-reducer.js [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+const INITIAL_STATE = {
+  products:
+    // JSON.parse(localStorage.getItem("products")).products ||
+    productsInitialState,
+  cart: [],
+  currentItem: null,
+};
+
+const shopReducer = (state = INITIAL_STATE, action) => {
+  switch (action.type) {
```
Comment (Tymofiev) [inline src/redux/Shopping/shopping-reducer.js:124]: create func for every case action

File: src/redux/Shopping/shopping-reducer.js (hunk 8 above)
Comment (Tymofiev) [reply src/redux/Shopping/shopping-reducer.js]: +1

File: src/redux/Shopping/shopping-reducer.js [hunk 31]
```diff
@@ -0,0 +1,136 @@
...
+      };
+      localStorage.setItem("products", JSON.stringify(stateCopyOnAddToCart));
+      return stateCopyOnAddToCart;
+
+    case actionTypes.REMOVE_FROM_CART:
+      let stateCopyOnRemoveFromCart = {
+        ...state,
+        cart: state.cart.filter((item) => item.id !== action.payload.id),
+      };
+      localStorage.setItem(
```
Comment (Tymofiev) [inline src/redux/Shopping/shopping-reducer.js]: ideally actions should not do anything except changing state, create wrapper - service

File: src/services/authentication.service.js [hunk 32]
```diff
@@ -0,0 +1,40 @@
+import { BehaviorSubject } from "rxjs";
+
+import { handleResponse } from "../helpers/handle-response";
+
+const currentUserSubject = new BehaviorSubject(
```
Comment (Tymofiev) [inline src/services/authentication.service.js:5]: using angular style?

File: src/services/authentication.service.js [hunk 33]
```diff
@@ -0,0 +1,40 @@
...
+    headers: { "Content-Type": "application/json" },
+    body: JSON.stringify({ username, password })
+  };
+
+  return fetch(`/users/authenticate`, requestOptions)
+    .then(handleResponse)
+    .then(user => {
+      // store user details and jwt token in local storage to keep user logged in between page refreshes
+      localStorage.setItem("currentUser", JSON.stringify(user));
+      currentUserSubject.next(user);
```
Comment (Tymofiev) [inline src/services/authentication.service.js:21]: not bad =)

File: src/services/authentication.service.js [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+const logout = () => {
+  // remove user from local storage to log user out
+  localStorage.removeItem("currentUser");
+  currentUserSubject.next(null);
+};
+
+export const authenticationService = {
+  login,
+  logout,
+  currentUser: currentUserSubject.asObservable(),
```
Comment (Tymofiev) [inline src/services/authentication.service.js:36]: like a BOSS)

File: src/components/Admin/ProductItem.js [hunk 35]
```diff
@@ -0,0 +1,90 @@
...
+  deleteProduct() {
+    const { id } = this.props.product;
+    this.props.deleteProduct(id);
+  }
+  editProduct() {
+    this.setState((prevState, props) => ({
+      isEdit: !prevState.isEdit,
+    }));
+  }
+  editProductSubmit() {
```
Comment (Tymofiev) [inline src/components/Admin/ProductItem.js:57]: needs more different and actual naming

**Review by Tymofiev — CHANGES_REQUESTED** [review]

**Pull request closed without merging**